lib/ansible/modules/cloud/docker/docker_node.py

```diff
@@ -0,0 +1,290 @@
+#!/usr/bin/python
+#
+# (c) 2019 Piotr Wojciechowski <piotr@it-playground.pl>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: docker_node
+short_description: Manage Docker Swarm node
+version_added: "2.8"
+description:
+    - Manages the Docker nodes via Swarm Manager
+    - Change node role
+    - Change node availability
+    - Change or remove node labels
+options:
+    state:
+        description:
+            - Set to C(update) to update node labels, availability and role
+        required: true
+        choices:
+          - update
+    name:
+        description:
+            - Swarm name of the node.
+            - Used with I(state=update).
+    labels:
+        description: User-defined key/value metadata.
+    labels_state:
+        description:
+            - Defines the operation on the lables assigned to node.
+            - Set to C(merge) to merge provided labels with already assigned ones. It will update existing keys.
+              The I(labels) must be specified.
+            - Set to C(replace) to replace assigned labels with provided ones. The I(labels) must be specified.
+            - Set to C(remove) to remove all labels assigned to node.
+        choices:
+          - merge
+          - replace
+          - remove
+        default: 'merge'
+    availability:
+        description: Node availability status to assign
+        choices:
+          - active
+          - pause
+          - drain
+    role:
+        description: Node role to assign
```

Comment:
You might want to add a sentence that the configured value will not be updated if `availability` is not specified.

Reply:
The `availability` is always specified, if not directly in playbook then the current node availability is used for API call.

Reply:
Why is that? Doesn't that exactly mean that the value is not updated when it is not specified?

Reply:
The `availability` and `role` always have to be included in API call, if either is missing then API call returns `400`. The Docker documentation and python `docker` module documentation and comments in the code are a little misleading so I tested it by myself

If you just call `/update` with the following body

```
{
  "Availability": "drain"
}
```
the response is APIError 400 together with error description
```
{
    "message": "invalid Role: \"\""
}
```

If you specify just the availability
```
{
  "Role": "worker"
}
```
the APIError is 400 and message
```
{
    "message": "invalid Availability: \"\""
}
```

The body of `/update` POST must at least include `Role` and `Availability` (and `Version` as a key in a call to be precise). This is checked on Docker side even if you are not updating those attributes. Only the `Label` is cleared to empty object `{}` and 'Name` removed if missing from POST request body.

In such case, if any of those two parameters are missing in Ansible module use in a playbook then values assigned to the node will be provided unchanged for the API call, this way you don't have to provide both of them in the playbook every time.

Reply:
So if you use the current values (returned by inspect), these values are not changed. That's what I meant in the original sentence that should be added (and that's usually the only thing the user cares about).

Reply:
Ah ok, I connected it with the identical comment to line 56 above and my understanding was that you want `availability` to be required parameter somehow